xarray/core/parallel.py

```diff
@@ -201,20 +231,34 @@ def map_blocks(
         * time     (time) object 1990-01-31 00:00:00 ... 1991-12-31 00:00:00
     """
 
-    def _wrapper(func, obj, to_array, args, kwargs):
+    def _wrapper(func, obj, to_array, args, kwargs, expected):
+        check_shapes = dict(obj.dims)
+        check_shapes.update(expected["shapes"])
+
         if to_array:
             obj = dataset_to_dataarray(obj)
 
         result = func(obj, *args, **kwargs)
 
+        # check all dims are present
+        missing_dimensions = set(expected["shapes"]) - set(result.sizes)
+        if missing_dimensions:
+            raise ValueError(
+                f"Dimensions {missing_dimensions} missing on returned object."
+            )
+
+        # check that index lengths are as expected
         for name, index in result.indexes.items():
-            if name in obj.indexes:
-                if len(index) != len(obj.indexes[name]):
+            if name in check_shapes:
+                if len(index) != check_shapes[name]:
                     raise ValueError(
-                        "Length of the %r dimension has changed. This is not allowed."
-                        % name
+                        f"Received dimension {name!r} of length {len(index)}. Expected length {check_shapes[name]}."
                     )
 
+        check_result_variables(result, expected, "coords")
+        if isinstance(result, Dataset):
+            check_result_variables(result, expected, "data_vars")
+
```

Comment:
I think the last thing that probably needs to be checked here are indexes on the returned object. We should verify those are as expected.